.github/workflows/firebase-hosting-pull-request.yml

```diff
@@ -1,22 +1,81 @@
-# This file was auto-generated by the Firebase CLI
-# https://github.com/firebase/firebase-tools
-
 name: Deploy to Firebase Hosting on PR
 on: pull_request
 permissions:
   checks: write
   contents: read
   pull-requests: write
 jobs:
-  build_and_preview:
+  code_check:
+    name: Code check
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+
+      - uses: actions/setup-node@v4
+        with:
+          node-version: 18
+          cache: "npm"
+          cache-dependency-path: |
+            hosting/package-lock.json
+
+      - name: Install global NPM tools
+        run: |
+          npm install -g firebase-tools@latest
+
+      - name: Cache hosting build
+        uses: actions/cache@v4
+        id: cache-hosting
+        with:
+          key: ${CI_COMMIT_REF_SLUG}
+          path: |
+            ./hosting/node_modules/
+            ./hosting/.next/cache/
+
+      - name: Install dependency
+        run: npm install
+        working-directory: ./hosting
+
+      - name: Linting files
+        run: npm run lint --no-fix
+        working-directory: ./hosting
+
+  build:
+    name: Build
+    needs: code_check
     if: ${{ github.event.pull_request.head.repo.full_name == github.repository }}
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v4
-      - run: npm ci && npm run build      
+
+      - uses: actions/setup-node@v4
+        with:
+          node-version: 18
+          cache: "npm"
+          cache-dependency-path: |
+            hosting/package-lock.json
+
+      - name: Install global NPM tools
+        run: |
+          npm install -g firebase-tools@latest
```

Comment:
Why we need firebase-tools to be installed?

Reply:
Good point. Old habit. We're using hosting action for deployment and don't need this. Thanks, I'll remove it.

Reply:
Done ✅